tests/test_bootloader.py

```diff
@@ -0,0 +1,362 @@
+    ==========
+    seconds : int
+      Number of seconds before Exception is raised.
+    """
+    signal.signal(signal.SIGALRM, timeout_handler)
+    signal.alarm(seconds)
+
+  def __enter__(self):
+    return self
+
+  def __exit__(self, *args):
+    self.cancel()
+
+  def cancel(self):
+    """ Cancel scheduled Exception. """
+    signal.alarm(0)
+
+def set_known_state(port):
+  """
+  Set Piksi into a known state (STM / NAP firmware).
+
+  Parameters
+  ==========
+  port : string
+    Filepath of virtual com port attached to Piksi.
+  """
+  if VERBOSE: print "Setting up Piksi at '%s' to known state" % port
+  with serial_link.get_driver(use_ftdi=False, port=port) as driver:
+    with Handler(driver.read, driver.write) as link:
+      link.start()
+
+      # Wait until we receive a heartbeat or bootloader handshake so we
+      # know what state Piksi is in.
+      if VERBOSE: print "  Waiting for Heartbeat or Bootloader Handshake"
+      with Bootloader(link) as piksi_bootloader:
+        with Heartbeat(link) as heartbeat:
+          # Throw an exception if a heartbeat or handshake
+          # is not received for 5 seconds.
+          with Timeout(5) as timeout:
+            while not heartbeat.received and not piksi_bootloader.handshake_received:
+              time.sleep(0.1)
+          if VERBOSE: print "  Received Heartbeat or Bootloader Handshake"
+          # If Piksi is in the application, reset it into the bootloader.
+          if heartbeat.received:
+            if VERBOSE: print "  Resetting Piksi"
+            link.send(SBP_MSG_RESET, "")
+
+      with Bootloader(link) as piksi_bootloader:
+        # Set Piksi into bootloader mode.
+        with Timeout(10) as timeout:
+          piksi_bootloader.wait_for_handshake()
+        piksi_bootloader.reply_handshake()
+        if VERBOSE: print "  Received bootloader handshake"
+
+        with flash.Flash(link, flash_type="STM",
+                 sbp_version=piksi_bootloader.sbp_version) as piksi_flash:
+          # Erase entire STM flash (except bootloader).
+          if VERBOSE: print "  Erasing STM"
+          with Timeout(30) as timeout:
+            for s in range(1,12):
+              piksi_flash.erase_sector(s)
+          # Write STM firmware.
+          if VERBOSE: print "  Programming STM"
+          with Timeout(100) as timeout:
+            piksi_flash.write_ihx(STM_FW, erase=False)
+
+        with flash.Flash(link, flash_type="M25",
+                 sbp_version=piksi_bootloader.sbp_version) as piksi_flash:
+          # Write NAP hexfile.
+          if VERBOSE: print "  Programming NAP"
+          with Timeout(250) as timeout:
+            piksi_flash.write_ihx(NAP_FW)
+
+        # Jump to the application firmware.
+        if VERBOSE: print "  Jumping to application"
+        piksi_bootloader.jump_to_app()
+
+        if VERBOSE: print ""
+
+class TestBootloader(unittest.TestCase):
+  """
+  Piksi bootloader tests.
+  """
+
+  @classmethod
+  def setUpClass(self):
```

Comment:
This is a method inherited from unittest.TestCase so unfortunately I can't do much about it.